### ISSUE TITLE
Vehicle cannons and machine guns can be fast reloaded with an exploit

### ISSUE THREAD

**Issue by AndrewTheel** [DarklightGames/DarkestHour#687]
Shell reload time can be reduced to 1-2 seconds on any tank (very noticeable on Soviet tanks) with an exploit.  For details please message me on slack.

**Comment by cmbasnett** [comment]
@AndrewTheel  I think the point of this is to lay the details out *here*.

**Comment by AndrewTheel** [comment]
Basically just switch back and forth for each "stage" of the reload and it takes 1-2 seconds to reload any cannon.

**Comment by AndrewTheel** [comment]
I did use this exploit the other day just to confirm that it actually works and gives a major advantage.  It does and is completely broken for some tanks, I could fire about every second and a half.

**Comment by AndrewTheel** [comment]
It works for hull MGs as well

**Comment by handsm** [comment]
I know what causes this and have written a fix. Just testing out a couple of possibilities for the best result. Will commit soon.

**Comment by cmbasnett** [comment]
What's the status on this one? If this gets fixed we could merge it into the live build sooner rather than later.

**Comment by AndrewTheel** [comment]
Yea this exploit is ruining the game for me (because now I can't help but use it).

**Comment by handsm** [comment]
Here's a write up, prior to commit. I've testes it as much as I can and it all works. I need to separate a lot of stuff, but will get that done by tomorrow and commit.

The problem is that when a reload timer is called it assumes a reload stage has just completed, so it always moves the reload state on to the next stage. When a player enters a vehicle weapon position - in this case having exited and re-entered immediately - the server passes the current reload state to the player's client. And if the weapon needs to start or resume reloading, an almost immediate reload timer is always called to kick that process off, and the timer moves it on another stage.

There is a clear difference between a timer call to start a new reload and one that has been called after a reload stage finishes. But the timer always moves on a stage. For a newly started reload this is worked around by starting with a dummy 'empty' reload stage. When a new reload starts, the state is set to 'empty' but almost immediately the timer is called and it moves it on to reloading part 1. Really this extra empty stage is doing the same job as a bool flag and could be thought of as bStartingNewReload, where the bool would prevent the timer advancing a stage on the first timer call and then be cleared.

With DH's pause/resume reload system, including for hull and coaxial MGs, a resumed reload needs to act like a new reload by not advancing the reload stage, otherwise it can be exploited like this. So the fix is pretty simple and just means adding a bool flag so the timer knows that a paused reload is being resumed and it should not advance a stage.

It actually results in the functionality being a bit cleaner and simpler, because the new bool can also flag a new reload is starting, as well as flagging a resumed reload - the required effect is the same. So I've called it bNewOrResumedReload and it deprecates the dummy 'empty stage'. Very conveniently a new or resumed is only ever started from one place, so the new bool gets set there, is picked up by the first timer call, prevents a stage advance and is cleared.

There's no added replication as the server and owning net client both run their reload timers independently. The server initiates this on the client, but after there the client takes care of it, including setting and clearing the bool flag - the processes being the same on server and client.

The reload pause is now a more active process, so when a player exits (or buttons up and can no longer reload an external MG) the reload is actively paused at that time. Before it was more passive and waited for the next timer call to detect there was no player in a valid loading position and pause the reload then.

The result of this change is if a player exits and re-enters during a reload, the timer will effectively repeat the last reload stage that never completed. This is better in principle: you must complete the reload stage or it doesn't count. The last reload stage sound gets repeated. The only small drawback is the last reload sound doesn't stop when you exit (same as before), but I don't think it's possible in the UT2004 engine to stop a sound playing once it's started.

**Comment by handsm** [comment]
Fixed in commit https://github.com/DarklightGames/DarkestHour/commit/a1d499d9ceeb72d39fb92129a5e670389e153837.

About to go away for 24 hours so have separated and committed this code in a bit of a rush, but believe all is ok. Will write up changes here tomorrow afternoon.

**Comment by cmbasnett** [comment]
We can give this a proper multiplayer test, then merge it over to the live build if it all checks out. The explanation sounds solid though, good work.

**Comment by cmbasnett** [comment]
Gave this a thorough multiplayer (dedicated server) test on my local machine and it all seems to check out. Will merge this into live!

**Comment by handsm** [comment]
An overview of the reload process and the changes I made:

All reload stuff starts with the AttemptReload() function, or the very similar AttemptAltReload() for a coaxial MG. This is called at various trigger points when the weapon is not loaded: fire cannon, fire last round in a magazine weapon, player enters, player moves to valid loading position, resupply ammo when have none, player does manual cannon reload, player switches off manual reloading. In addition for a coaxial MG when a cannon reload completes (both can't reload together and cannon takes priority).

AttemptReload() is called on server and client. Normally the server initiates the process and causes the owning client to run AttemptReload() using the replicated server-to-client function ClientSetReloadState(). The exception is if a player finishes transitioning to a new position and a reload is paused, where a net client triggers its own AttemptReload() to check if it should resume reloading.

AttemptReload() remains not significantly changed and runs the checks needed to make sure a reload can start or resume. If it can it calls a function that was called StartReloadTimer(), which I have now renamed to StartReload(). This is the only place a reload timer is initiated, kicking off the new/resumed reload process. The critical change is that it now sets bNewOrResumedReload, which is picked up in the first timer call (and cleared there) and stops the timer from advancing to the next reload stage.

As the new bool flag applies to a new reload as well as a resumed reload, I have deprecated the now pointless 'empty' reload stage. I have also renamed the stages from RL_ReloadedPart1 etc to RL_ReloadingPart1 etc. This isn't just semantics; the old name was misleading. In part/stage 1 the weapon is not reloadED part 1 - the point of this issue and fix is that in this state no reload stage has actually been completed; we are doing reloadING part 1 and nothing has been completed. Unless the player stays and completes this stage, we don't move to a 25% reloaded state. At the other end, in stage/part 4, we have completed 75% of the reload  and are in the process of completing the last 25%.

These changes mean the reload stage increment and the match to the ReloadStages array index is cleaner, simpler and more obvious, without having to re-jig the order and apply adjustments to the array index in some places, as before.

The other small change in StartReload() is it now includes decrementing the ammo/mag at the start of a new reload, which was previously included in AttemptReload(). Makes subclassing easier, strips down the longer AttemptReload() a little, and we end up with a StartReload() function that does what is says. AttemptReload() does the checks and calls the appropriate function; StartReload() does the work.

When a reload needs to be paused, this now happens actively instead of letting it run on to the next reload timer and letting that detect the lack of player in a valid loading position (or whatever). When a pause is triggered it now calls a new function called PauseReload(). Apart from setting the obvious pause bool flag, it now actively stops the timer. Important note here: a coaxial MG needs to do an additional check that the cannon isn't reloading before it stops the timer, otherwise a paused coax reload would also kill the cannon reload. PauseReload() also allows easy subclassing and I have used it to move the functionality that pauses an MG HUD overlay reload, which was previously tacked on to the end of Timer().

We end up with three functions - AttemptReload(), StartReload() and PauseReload() - that do exactly what it says on the tin and are easily subclassed, and reload stage names that are easy to follow and accurately describe the stage. A minor point but RL_Part1 etc would probably better be renamed to RL_Stage1 for slightly better clarity/directness.
